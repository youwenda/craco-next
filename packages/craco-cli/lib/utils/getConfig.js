const fs = require('fs');
const path = require('path');
const nodeLibs = require('node-libs-browser');
const webpack = require('webpack');
const WebpackBar = require('webpackbar');
const paths = require('react-scripts/config/paths');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const { configOverrides } = require('./paths');
const FilterCSSConflictingWarning = require('../plugins/FilterCSSConflictingWarning');
const ConsoleBrowserUrlPlugin = require('../plugins/ConsoleBrowserUrlPlugin');
const {
  override,
  addBundleVisualizer,
  addLessLoader,
  addFileLoader,
  addSassLoader,
  addCssLoader,
  addPlugin,
  addMultipleEntries,
  addWebpackAlias,
  addSourceMapLoader,
  removePlugin,
  removeModuleScopePlugin,
  fixBabelImports,
  mergeOutput,
  mergeOptimization,
  mergeWebpackResolveFallback,
  useBabelRc,
  setWebpackStats,
} = require('customize-cra-extra');

function getUserConfig() {
  let userConfig = {};
  if (fs.existsSync(configOverrides)) {
    userConfig = require(configOverrides);
  }

  const webpack =
    typeof userConfig === 'function' ? userConfig : userConfig.webpack || ((config) => config);
  const devServer =
    userConfig.devServer ||
    userConfig.devserver ||
    ((configFunction) => (proxy, allowedHost) => configFunction(proxy, allowedHost));
  const jest = userConfig.jest || ((config) => config);
  const paths = userConfig.paths || ((paths) => paths);

  return {
    webpack,
    devServer,
    jest,
    paths,
  };
}

function getConfig() {
  const userConfig = getUserConfig();
  return {
    /**
     * [webpack description]
     *
     * @param   {Object}  config  [CRA's webpack config]
     * @param   {String}  env     [process.env.NODE_ENV 'development' || 'production']
     *
     * @return  {Object}          [return new webpack config]
     */
    webpack(config, env) {
      const isProduction = env === 'production';
      const isDevelopment = env === 'development';

      const {
        DISABLE_NODE_POLYFILLS,
        GENERATE_BUNDLE_REPORT,
        TS_CHECKER,
        NO_VENDORS_DLL,
        USE_STYLE_LOADER,
      } = process.env;
      const disableNodePolyfills = DISABLE_NODE_POLYFILLS !== 'false';

      const webpackConfig = override(
        // Entry
        addMultipleEntries(
          [
            {
              entry: 'index',
              template: paths.appHtml,
            },
          ],
          paths.appSrc,
          true /* internal used only */
        ),
        // Output
        mergeOutput({
          filename: '[name].js',
          // There are also additional JS chunk files if you use code splitting.
          chunkFilename: `chunks/[name]${isDevelopment ? '' : '.[contenthash:8]'}.js`,
          assetModuleFilename: 'media/[name].[hash][ext]',
        }),

        // Loaders
        addLessLoader(),
        addSassLoader(),
        addCssLoader(),
        addSourceMapLoader(),
        addFileLoader(),
        useBabelRc(), // eslint-disable-line
        // Plugins
        // removePlugin(GenerateSW),
        removeModuleScopePlugin(),
        removePlugin('WebpackManifestPlugin'),
        TS_CHECKER === 'false' && removePlugin('ForkTsCheckerWarningWebpackPlugin'),
        TS_CHECKER === 'false' && removePlugin('ForkTsCheckerWebpackPlugin'),

        isProduction && removePlugin(MiniCssExtractPlugin),
        isProduction && GENERATE_BUNDLE_REPORT && addBundleVisualizer(),
        //@note 适配customize-cra-extra 1.0.1-alpha.1版本mini-css-extract-plugin一直存在的情况
        USE_STYLE_LOADER !== 'true' &&
          addPlugin(
            new MiniCssExtractPlugin({
              // Options similar to the same options in webpackOptions.output
              // both options are optional
              // filename: 'static/css/[name].[contenthash:8].css',
              // chunkFilename: 'static/css/[name].[contenthash:8].chunk.css',
              filename: '[name].css',
              chunkFilename: `chunks/[name]${isDevelopment ? '' : '.[contenthash:8]'}.css`,
            })
          ),
        NO_VENDORS_DLL !== 'true' &&
          addPlugin(
            new webpack.DllReferencePlugin({
              context: paths.appPath,
              manifest: path.resolve(
                isDevelopment ? paths.appPublic : paths.appBuild,
                './vendors/vendors-manifest.json'
              ),
            })
          ),
        addPlugin(new WebpackBar()),
        addPlugin(new FilterCSSConflictingWarning()),
        !disableNodePolyfills &&
          addPlugin(
            new webpack.ProvidePlugin({
              process: nodeLibs['process'],
            })
          ),
        !disableNodePolyfills &&
          addPlugin(
            // ref: https://github.com/umijs/umi/issues/6914
            new webpack.ProvidePlugin({
              Buffer: ['buffer', 'Buffer'],
            })
          ),
        isDevelopment && addPlugin(new ConsoleBrowserUrlPlugin()),

        // babel import plugin
        fixBabelImports('lodash', {
          libraryDirectory: '',
          camel2DashComponentName: false,
        }),

        // optimization
        mergeOptimization({
          splitChunks: {
            chunks: 'async',
            name: false,
          },
          runtimeChunk: !isProduction,
        }),

        // fallback node
        !disableNodePolyfills &&
          mergeWebpackResolveFallback(
            Object.assign(
              Object.keys(nodeLibs).reduce((memo, key) => {
                if (nodeLibs[key]) {
                  memo[key] = nodeLibs[key];
                } else {
                  memo[key] = false;
                }
                return memo;
              }, {}),
              {
                http: false,
                https: false,
              }
            )
          ),
        // stats
        setWebpackStats('errors-only'),

        addWebpackAlias({
          '@': paths.appSrc,
        })
      )(config, env);

      // 交由项目配置再次更改
      return userConfig.webpack(webpackConfig, env);
    },
    /**
     * [devServer description]
     *
     * @param   {Function}  configFunction  [CRA's webpackDevServer.config.js module.exports ]
     * @param   {String}  env             [process.env.NODE_ENV 'development' || 'production']
     *
     * @return  {Object}                  [return new webpack devServer config]
     */
    devServer(configFunction, env) {
      const devServer = function (proxy, allowedHost) {
        const { HTTPS_KEY, HTTPS_CERT } = process.env;

        if (HTTPS_KEY) {
          process.env.SSL_KEY_FILE = HTTPS_KEY;
        }
        if (HTTPS_CERT) {
          process.env.SSL_CERT_FILE = HTTPS_CERT;
        }
        // Create the default config by calling configFunction with the proxy/allowedHost parameters
        const config = configFunction(proxy, allowedHost);
        // Change the https certificate options to match your certificate, using the .env file to
        // set the file paths & passphrase.

        let onBeforeSetupMiddleware, onAfterSetupMiddleware;
        if (typeof config.onBeforeSetupMiddleware === 'function') {
          onBeforeSetupMiddleware = config.onBeforeSetupMiddleware;
          delete config.onBeforeSetupMiddleware;
        }
        if (typeof config.onAfterSetupMiddleware === 'function') {
          onAfterSetupMiddleware = config.onAfterSetupMiddleware;
          delete config.onAfterSetupMiddleware;
        }

        config.setupMiddlewares = (middlewares, devServer) => {
          if (onBeforeSetupMiddleware) {
            onBeforeSetupMiddleware(devServer);
          }

          if (onAfterSetupMiddleware) {
            onAfterSetupMiddleware(devServer);
          }
          return middlewares;
        };

        // Return your customised Webpack Development Server config.
        return config;
      };
      // 交由项目配置再次更改
      return userConfig.devServer(devServer, env);
    },
    paths(paths, env) {
      return userConfig.paths(paths, env);
    },
    jest(...args) {
      return userConfig.jest(...args);
    },
  };
}

module.exports = getConfig;
