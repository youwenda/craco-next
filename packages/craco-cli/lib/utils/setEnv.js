const httpsReg = /^(?:https?:)?\/\//;

// 设置默认值
Object.assign(process.env, {
  // disable create inline runtime chunk
  INLINE_RUNTIME_CHUNK: 'false',
  // disable eslint plugin
  DISABLE_ESLINT_PLUGIN: 'true',
  // disable node polyfills
  DISABLE_NODE_POLYFILLS: 'false',
  // enable generate sourceMap
  GENERATE_SOURCEMAP: 'true',
  // disable clear console
  CLEAR_CONSOLE: 'false',
  // disable forkTs
  TS_CHECKER: 'false',
  // disable tsc
  TS_COMPILER: 'false',
  // disable verifyTypeScriptSetup
  TS_AUTO_SETUP: 'false',
  // disable auto opening browsers
  BROWSER: 'none',
  // disable auto opening browsers，Isomorphism from BROWSER=none
  AUTO_OPEN_BROWSER: 'false',
});

// 兼容 .rmxrc env 之前，需要同步react-scripts的env逻辑
require('react-scripts/config/env');

function setEnv() {
  // 订正可能存在异常数据设置HOST
  if (typeof process.env.HOST === 'string') {
    process.env.HOST = process.env.HOST.replace(httpsReg, '');
  }
}

module.exports = setEnv;
