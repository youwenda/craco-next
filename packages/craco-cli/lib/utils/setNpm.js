const TerserPlugin = require('terser-webpack-plugin');

require('react-dev-utils/clearConsole');
require.cache[require.resolve('react-dev-utils/clearConsole')].exports = function clearConsole() {
  const { CLEAR_CONSOLE = 'false' } = process.env;
  if (CLEAR_CONSOLE !== 'false') {
    process.stdout.write(process.platform === 'win32' ? '\x1B[2J\x1B[0f' : '\x1B[2J\x1B[3J\x1B[H');
  }
};

// 初次启动时检测存在可能的接口时间会很长
// require('react-dev-utils/getProcessForPort');
// require.cache[require.resolve('react-dev-utils/getProcessForPort')].exports =
//   function getProcessForPort() {};

const { createCompiler } = require('react-dev-utils/WebpackDevServerUtils');
const { TS_CHECKER } = process.env;
if (!TS_CHECKER || TS_CHECKER === 'false') {
  // 强制不处理typescript
  require.cache[require.resolve('react-dev-utils/WebpackDevServerUtils')].exports.createCompiler = (
    options = {}
  ) => {
    return createCompiler({
      ...options,
      useTypeScript: false,
    });
  };
}

const verifyTypeScriptSetup = require('react-scripts/scripts/utils/verifyTypeScriptSetup');
require.cache[require.resolve('react-scripts/scripts/utils/verifyTypeScriptSetup')].exports =
  function () {
    const { TS_AUTO_SETUP = 'true' } = process.env;
    if (TS_AUTO_SETUP !== 'true') {
      return;
    }
    verifyTypeScriptSetup();
  };

require.cache[
  require.resolve('terser-webpack-plugin')
].exports = class TerserPluginOverridden extends TerserPlugin {
  constructor(options) {
    options.extractComments = false;
    super(options);
  }
};

const origOpenBrowser = require('react-dev-utils/openBrowser');
require.cache[require.resolve('react-dev-utils/openBrowser')].exports = function openBrowser(
  ...args
) {
  const { AUTO_OPEN_BROWSER = 'false' } = process.env;
  if (AUTO_OPEN_BROWSER !== 'false') {
    return origOpenBrowser(...args);
  } else {
    process.env.AUTO_OPEN_BROWSER_URL = args[0];
  }
};
