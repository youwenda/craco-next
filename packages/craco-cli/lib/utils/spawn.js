const spawn = require('cross-spawn');

function spawnPromise(command, args, options = { stdio: 'inherit' }) {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, options);
    child
      .on('error', (err) => {
        reject(err);
      })
      .on('close', (code) => {
        if (code !== 0) {
          reject({
            command: `${command} ${args.join(' ')}`,
          });
          return;
        }
        resolve();
      });
  });
}

function spawnSync(command, args, options = { stdio: 'inherit' }) {
  return spawn.sync(command, args, options);
}

exports.spawnPromise = spawnPromise;
exports.spawnSync = spawnSync;
