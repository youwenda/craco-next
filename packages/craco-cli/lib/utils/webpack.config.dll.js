const path = require('path');
const webpack = require('webpack');
const { CleanWebpackPlugin } = require('clean-webpack-plugin');
const paths = require('react-scripts/config/paths');
const appPackage = require(paths.appPackageJson);
const library = 'vendors';
const isProduction = process.env.NODE_ENV === 'production';
const shouldUseSourceMap = process.env.GENERATE_SOURCEMAP !== 'false';
const vendorsExtra = [];

let vendorsDll = Array.from(
  new Set([
    'es6-promise',
    'classnames',
    'history',
    'react',
    'react-dom',
    'react-redux',
    'react-router-dom',
    'redux',
    ...vendorsExtra,
  ])
).filter((vendor) => appPackage.dependencies[vendor]);
let outputPath = path.resolve(paths.appPublic, library);

if (isProduction) {
  outputPath = path.resolve(paths.appBuild, library);
}

const config = {
  mode: isProduction ? 'production' : 'development',
  entry: {
    vendorsDll,
  },
  output: {
    filename: 'vendors.dll.js',
    path: outputPath,
    library,
  },
  plugins: [
    new CleanWebpackPlugin(),
    new webpack.DllPlugin({
      path: path.resolve(outputPath, './vendors-manifest.json'),
      // This must match the output.library option above
      name: library,
    }),
  ],
  devtool: shouldUseSourceMap ? 'source-map' : false,
};

module.exports = config;
