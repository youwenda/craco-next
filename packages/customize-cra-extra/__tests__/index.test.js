require('../../craco-cli/lib/utils/setEnv');
require('../../craco-cli/lib/utils/setNpm');
const { addLoaders } = require('../lib');
const { removePlugin } = require('../lib');
const { cleanPlugin } = require('../lib');
const { addPlugins } = require('../lib');
const {
  addBabelIncludes,
  getBabelLoader,
  removeInternalBabelPlugin,
  addSourceMapLoader,
} = require('../lib');
const { getConfig } = require('./config');

const ForkTsCheckerWebpackPlugin =
  process.env.TSC_COMPILE_ON_ERROR === 'true'
    ? require('react-dev-utils/ForkTsCheckerWarningWebpackPlugin')
    : require('react-dev-utils/ForkTsCheckerWebpackPlugin');

describe('addBabelIncludes', () => {
  test('add string in src babel', () => {
    const developmentConfig = getConfig();
    expect(getBabelLoader(addBabelIncludes('lib')(developmentConfig))).toMatchSnapshot();
  });
  test('add array in src babel', () => {
    const developmentConfig = getConfig();
    expect(getBabelLoader(addBabelIncludes(['lib', 'lib2'])(developmentConfig))).toMatchSnapshot();
  });
  test('add array in node_modules babel', () => {
    const developmentConfig = getConfig();
    expect(
      getBabelLoader(addBabelIncludes(['lib', 'lib2'], true)(developmentConfig), true)
    ).toMatchSnapshot();
  });
});

describe('removeInternalBabelPlugin', () => {
  test('remove react-refresh', () => {
    const developmentConfig = getConfig();
    expect(
      getBabelLoader(removeInternalBabelPlugin('react-refresh')(developmentConfig))
    ).toMatchSnapshot();
  });
});

describe('addLoaders', () => {
  test("addLoaders [{test:'.jsx',loader: 'babel-loader'}, {test: 'jsx2', use: [{ loader: 'babel-loader' }]}]", () => {
    const developmentConfig = getConfig();
    expect(
      addLoaders([
        { test: '.jsx', loader: 'babel-loader' },
        { test: 'jsx2', use: [{ loader: 'babel-loader' }] },
      ])(developmentConfig).module.rules[1].oneOf.slice(-3)
    ).toMatchSnapshot();
  });
});

describe('addPlugins', () => {
  test('add one plugin', () => {
    const developmentConfig = getConfig();
    const plugin = function Plugin() {};
    expect(addPlugins([new plugin()])(developmentConfig).plugins).toMatchSnapshot();
  });
});

describe('removelugins', () => {
  test('remove ESLintWebpackPlugin plugin', () => {
    const developmentConfig = getConfig();
    expect(removePlugin('ESLintWebpackPlugin')(developmentConfig).plugins).toMatchSnapshot();
  });
  test('clean ForkTsCheckerWebpackPlugin plugin', () => {
    const developmentConfig = getConfig();
    expect(cleanPlugin(ForkTsCheckerWebpackPlugin)(developmentConfig).plugins).toMatchSnapshot();
  });
});

describe('addSourceMapLoader', () => {
  test('changeSourceMapLoader', () => {
    const developmentConfig = getConfig();
    expect(addSourceMapLoader()(developmentConfig).module.rules[0]).toMatchSnapshot();
  });
});
