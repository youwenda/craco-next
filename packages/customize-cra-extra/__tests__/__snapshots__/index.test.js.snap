// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addBabelIncludes add array in node_modules babel 1`] = `
Object {
  "exclude": /@babel\\(\\?:\\\\/\\|\\\\\\\\\\{1,2\\}\\)runtime/,
  "include": Array [
    "lib",
    "lib2",
  ],
  "loader": "craco-next/node_modules/babel-loader/lib/index.js",
  "options": Object {
    "babelrc": false,
    "cacheCompression": false,
    "cacheDirectory": true,
    "cacheIdentifier": "development:babel-plugin-named-asset-import@0.3.8:babel-preset-react-app@10.0.1:react-dev-utils@12.0.1:react-scripts@5.0.1",
    "compact": false,
    "configFile": false,
    "inputSourceMap": true,
    "presets": Array [
      Array [
        "craco-next/node_modules/babel-preset-react-app/dependencies.js",
        Object {
          "helpers": true,
        },
      ],
    ],
    "sourceMaps": true,
  },
  "test": /\\\\\\.\\(js\\|mjs\\)\\$/,
}
`;

exports[`addBabelIncludes add array in src babel 1`] = `
Object {
  "include": Array [
    "craco-next/packages/customize-cra-extra/src",
    "lib",
    "lib2",
  ],
  "loader": "craco-next/node_modules/babel-loader/lib/index.js",
  "options": Object {
    "babelrc": false,
    "cacheCompression": false,
    "cacheDirectory": true,
    "cacheIdentifier": "development:babel-plugin-named-asset-import@0.3.8:babel-preset-react-app@10.0.1:react-dev-utils@12.0.1:react-scripts@5.0.1",
    "compact": false,
    "configFile": false,
    "customize": "craco-next/node_modules/babel-preset-react-app/webpack-overrides.js",
    "plugins": Array [
      "craco-next/node_modules/react-refresh/babel.js",
    ],
    "presets": Array [
      Array [
        "craco-next/node_modules/babel-preset-react-app/index.js",
        Object {
          "runtime": "classic",
        },
      ],
    ],
  },
  "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
}
`;

exports[`addBabelIncludes add string in src babel 1`] = `
Object {
  "include": Array [
    "craco-next/packages/customize-cra-extra/src",
    "lib",
  ],
  "loader": "craco-next/node_modules/babel-loader/lib/index.js",
  "options": Object {
    "babelrc": false,
    "cacheCompression": false,
    "cacheDirectory": true,
    "cacheIdentifier": "development:babel-plugin-named-asset-import@0.3.8:babel-preset-react-app@10.0.1:react-dev-utils@12.0.1:react-scripts@5.0.1",
    "compact": false,
    "configFile": false,
    "customize": "craco-next/node_modules/babel-preset-react-app/webpack-overrides.js",
    "plugins": Array [
      "craco-next/node_modules/react-refresh/babel.js",
    ],
    "presets": Array [
      Array [
        "craco-next/node_modules/babel-preset-react-app/index.js",
        Object {
          "runtime": "classic",
        },
      ],
    ],
  },
  "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
}
`;

exports[`addLoaders addLoaders [{test:'.jsx',loader: 'babel-loader'}, {test: 'jsx2', use: [{ loader: 'babel-loader' }]}] 1`] = `
Array [
  Object {
    "loader": "babel-loader",
    "test": ".jsx",
  },
  Object {
    "test": "jsx2",
    "use": Array [
      Object {
        "loader": "babel-loader",
      },
    ],
  },
  Object {
    "exclude": Array [
      /\\^\\$/,
      /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
      /\\\\\\.html\\$/,
      /\\\\\\.json\\$/,
    ],
    "type": "asset/resource",
  },
]
`;

exports[`addPlugins add one plugin 1`] = `
Array [
  HtmlWebpackPlugin {
    "userOptions": Object {
      "inject": true,
      "template": "craco-next/packages/customize-cra-extra/public/index.html",
    },
    "version": 5,
  },
  InterpolateHtmlPlugin {
    "htmlWebpackPlugin": [Function],
    "replacements": Object {
      "FAST_REFRESH": true,
      "NODE_ENV": "test",
      "PUBLIC_URL": "",
      "WDS_SOCKET_HOST": undefined,
      "WDS_SOCKET_PATH": undefined,
      "WDS_SOCKET_PORT": undefined,
    },
  },
  ModuleNotFoundPlugin {
    "appPath": "craco-next/packages/customize-cra-extra",
    "getRelativePath": [Function],
    "prettierError": [Function],
    "useYarnCommand": [Function],
    "yarnLockFile": undefined,
  },
  DefinePlugin {
    "definitions": Object {
      "process.env": Object {
        "FAST_REFRESH": "true",
        "NODE_ENV": "\\"test\\"",
        "PUBLIC_URL": "\\"\\"",
        "WDS_SOCKET_HOST": undefined,
        "WDS_SOCKET_PATH": undefined,
        "WDS_SOCKET_PORT": undefined,
      },
    },
  },
  ReactRefreshPlugin {
    "options": Object {
      "exclude": /node_modules/i,
      "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
      "overlay": false,
    },
  },
  CaseSensitivePathsPlugin {
    "fsOperations": 0,
    "logger": BufferedConsole {
      "Console": [Function],
      "_buffer": Array [],
      "_counters": Object {},
      "_groupDepth": 0,
      "_timers": Object {},
      "assert": [Function],
      "clear": [Function],
      "count": [Function],
      "countReset": [Function],
      "debug": [Function],
      "dir": [Function],
      "dirxml": [Function],
      "error": [Function],
      "group": [Function],
      "groupCollapsed": [Function],
      "groupEnd": [Function],
      "info": [Function],
      "log": [Function],
      "table": [Function],
      "time": [Function],
      "timeEnd": [Function],
      "timeLog": [Function],
      "trace": [Function],
      "warn": [Function],
    },
    "options": Object {},
    "pathCache": Map {},
    "primed": false,
  },
  WebpackManifestPlugin {
    "options": Object {
      "basePath": "",
      "fileName": "asset-manifest.json",
      "filter": null,
      "generate": [Function],
      "map": null,
      "publicPath": "/",
      "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
      "seed": undefined,
      "serialize": [Function],
      "sort": null,
      "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
      "useEntryKeys": false,
      "useLegacyEmit": false,
      "writeToFileEmit": false,
    },
  },
  IgnorePlugin {
    "checkIgnore": [Function],
    "options": Object {
      "contextRegExp": /moment\\$/,
      "resourceRegExp": /\\^\\\\\\.\\\\/locale\\$/,
    },
  },
  ForkTsCheckerWebpackPlugin {
    "options": Object {
      "async": true,
      "issue": Object {
        "exclude": Array [
          Object {
            "file": "**/src/**/__tests__/**",
          },
          Object {
            "file": "**/src/**/?(*.){spec|test}.*",
          },
          Object {
            "file": "**/src/setupProxy.*",
          },
          Object {
            "file": "**/src/setupTests.*",
          },
        ],
        "include": Array [
          Object {
            "file": "../**/src/**/*.{ts,tsx}",
          },
          Object {
            "file": "**/src/**/*.{ts,tsx}",
          },
        ],
      },
      "logger": Object {
        "infrastructure": "silent",
      },
      "typescript": Object {
        "configOverwrite": Object {
          "compilerOptions": Object {
            "declarationMap": false,
            "incremental": true,
            "inlineSourceMap": false,
            "noEmit": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "tsBuildInfoFile": "craco-next/packages/customize-cra-extra/node_modules/.cache/tsconfig.tsbuildinfo",
          },
        },
        "context": "craco-next/packages/customize-cra-extra",
        "diagnosticOptions": Object {
          "syntactic": true,
        },
        "mode": "write-references",
        "typescriptPath": "craco-next/node_modules/typescript/lib/typescript.js",
      },
    },
  },
  Plugin {},
]
`;

exports[`addSourceMapLoader changeSourceMapLoader 1`] = `
Object {
  "enforce": "pre",
  "exclude": /node_modules/,
  "loader": "craco-next/node_modules/source-map-loader/dist/cjs.js",
  "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\|css\\)\\$/,
}
`;

exports[`removeInternalBabelPlugin remove react-refresh 1`] = `
Object {
  "include": "craco-next/packages/customize-cra-extra/src",
  "loader": "craco-next/node_modules/babel-loader/lib/index.js",
  "options": Object {
    "babelrc": false,
    "cacheCompression": false,
    "cacheDirectory": true,
    "cacheIdentifier": "development:babel-plugin-named-asset-import@0.3.8:babel-preset-react-app@10.0.1:react-dev-utils@12.0.1:react-scripts@5.0.1",
    "compact": false,
    "configFile": false,
    "customize": "craco-next/node_modules/babel-preset-react-app/webpack-overrides.js",
    "plugins": Array [],
    "presets": Array [
      Array [
        "craco-next/node_modules/babel-preset-react-app/index.js",
        Object {
          "runtime": "classic",
        },
      ],
    ],
  },
  "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
}
`;

exports[`removelugins clean ForkTsCheckerWebpackPlugin plugin 1`] = `
Array [
  HtmlWebpackPlugin {
    "userOptions": Object {
      "inject": true,
      "template": "craco-next/packages/customize-cra-extra/public/index.html",
    },
    "version": 5,
  },
  InterpolateHtmlPlugin {
    "htmlWebpackPlugin": [Function],
    "replacements": Object {
      "FAST_REFRESH": true,
      "NODE_ENV": "test",
      "PUBLIC_URL": "",
      "WDS_SOCKET_HOST": undefined,
      "WDS_SOCKET_PATH": undefined,
      "WDS_SOCKET_PORT": undefined,
    },
  },
  ModuleNotFoundPlugin {
    "appPath": "craco-next/packages/customize-cra-extra",
    "getRelativePath": [Function],
    "prettierError": [Function],
    "useYarnCommand": [Function],
    "yarnLockFile": undefined,
  },
  DefinePlugin {
    "definitions": Object {
      "process.env": Object {
        "FAST_REFRESH": "true",
        "NODE_ENV": "\\"test\\"",
        "PUBLIC_URL": "\\"\\"",
        "WDS_SOCKET_HOST": undefined,
        "WDS_SOCKET_PATH": undefined,
        "WDS_SOCKET_PORT": undefined,
      },
    },
  },
  ReactRefreshPlugin {
    "options": Object {
      "exclude": /node_modules/i,
      "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
      "overlay": false,
    },
  },
  CaseSensitivePathsPlugin {
    "fsOperations": 0,
    "logger": BufferedConsole {
      "Console": [Function],
      "_buffer": Array [],
      "_counters": Object {},
      "_groupDepth": 0,
      "_timers": Object {},
      "assert": [Function],
      "clear": [Function],
      "count": [Function],
      "countReset": [Function],
      "debug": [Function],
      "dir": [Function],
      "dirxml": [Function],
      "error": [Function],
      "group": [Function],
      "groupCollapsed": [Function],
      "groupEnd": [Function],
      "info": [Function],
      "log": [Function],
      "table": [Function],
      "time": [Function],
      "timeEnd": [Function],
      "timeLog": [Function],
      "trace": [Function],
      "warn": [Function],
    },
    "options": Object {},
    "pathCache": Map {},
    "primed": false,
  },
  WebpackManifestPlugin {
    "options": Object {
      "basePath": "",
      "fileName": "asset-manifest.json",
      "filter": null,
      "generate": [Function],
      "map": null,
      "publicPath": "/",
      "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
      "seed": undefined,
      "serialize": [Function],
      "sort": null,
      "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
      "useEntryKeys": false,
      "useLegacyEmit": false,
      "writeToFileEmit": false,
    },
  },
  IgnorePlugin {
    "checkIgnore": [Function],
    "options": Object {
      "contextRegExp": /moment\\$/,
      "resourceRegExp": /\\^\\\\\\.\\\\/locale\\$/,
    },
  },
]
`;

exports[`removelugins remove ESLintWebpackPlugin plugin 1`] = `
Array [
  HtmlWebpackPlugin {
    "userOptions": Object {
      "inject": true,
      "template": "craco-next/packages/customize-cra-extra/public/index.html",
    },
    "version": 5,
  },
  InterpolateHtmlPlugin {
    "htmlWebpackPlugin": [Function],
    "replacements": Object {
      "FAST_REFRESH": true,
      "NODE_ENV": "test",
      "PUBLIC_URL": "",
      "WDS_SOCKET_HOST": undefined,
      "WDS_SOCKET_PATH": undefined,
      "WDS_SOCKET_PORT": undefined,
    },
  },
  ModuleNotFoundPlugin {
    "appPath": "craco-next/packages/customize-cra-extra",
    "getRelativePath": [Function],
    "prettierError": [Function],
    "useYarnCommand": [Function],
    "yarnLockFile": undefined,
  },
  DefinePlugin {
    "definitions": Object {
      "process.env": Object {
        "FAST_REFRESH": "true",
        "NODE_ENV": "\\"test\\"",
        "PUBLIC_URL": "\\"\\"",
        "WDS_SOCKET_HOST": undefined,
        "WDS_SOCKET_PATH": undefined,
        "WDS_SOCKET_PORT": undefined,
      },
    },
  },
  ReactRefreshPlugin {
    "options": Object {
      "exclude": /node_modules/i,
      "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
      "overlay": false,
    },
  },
  CaseSensitivePathsPlugin {
    "fsOperations": 0,
    "logger": BufferedConsole {
      "Console": [Function],
      "_buffer": Array [],
      "_counters": Object {},
      "_groupDepth": 0,
      "_timers": Object {},
      "assert": [Function],
      "clear": [Function],
      "count": [Function],
      "countReset": [Function],
      "debug": [Function],
      "dir": [Function],
      "dirxml": [Function],
      "error": [Function],
      "group": [Function],
      "groupCollapsed": [Function],
      "groupEnd": [Function],
      "info": [Function],
      "log": [Function],
      "table": [Function],
      "time": [Function],
      "timeEnd": [Function],
      "timeLog": [Function],
      "trace": [Function],
      "warn": [Function],
    },
    "options": Object {},
    "pathCache": Map {},
    "primed": false,
  },
  WebpackManifestPlugin {
    "options": Object {
      "basePath": "",
      "fileName": "asset-manifest.json",
      "filter": null,
      "generate": [Function],
      "map": null,
      "publicPath": "/",
      "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
      "seed": undefined,
      "serialize": [Function],
      "sort": null,
      "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
      "useEntryKeys": false,
      "useLegacyEmit": false,
      "writeToFileEmit": false,
    },
  },
  IgnorePlugin {
    "checkIgnore": [Function],
    "options": Object {
      "contextRegExp": /moment\\$/,
      "resourceRegExp": /\\^\\\\\\.\\\\/locale\\$/,
    },
  },
  ForkTsCheckerWebpackPlugin {
    "options": Object {
      "async": true,
      "issue": Object {
        "exclude": Array [
          Object {
            "file": "**/src/**/__tests__/**",
          },
          Object {
            "file": "**/src/**/?(*.){spec|test}.*",
          },
          Object {
            "file": "**/src/setupProxy.*",
          },
          Object {
            "file": "**/src/setupTests.*",
          },
        ],
        "include": Array [
          Object {
            "file": "../**/src/**/*.{ts,tsx}",
          },
          Object {
            "file": "**/src/**/*.{ts,tsx}",
          },
        ],
      },
      "logger": Object {
        "infrastructure": "silent",
      },
      "typescript": Object {
        "configOverwrite": Object {
          "compilerOptions": Object {
            "declarationMap": false,
            "incremental": true,
            "inlineSourceMap": false,
            "noEmit": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "tsBuildInfoFile": "craco-next/packages/customize-cra-extra/node_modules/.cache/tsconfig.tsbuildinfo",
          },
        },
        "context": "craco-next/packages/customize-cra-extra",
        "diagnosticOptions": Object {
          "syntactic": true,
        },
        "mode": "write-references",
        "typescriptPath": "craco-next/node_modules/typescript/lib/typescript.js",
      },
    },
  },
]
`;
